src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -842,6 +876,61 @@ private void CheckForComplete()
             }
         }
 
+        private Runspace GetRunspace(int taskId)
+        {
+            var runspaceName = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", PSTask.RunspaceName, taskId);
+
+            if (_runspacePool.TryDequeue(out Runspace runspace))
+            {
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    runspace.Name = runspaceName;
+                    return runspace;
+                }
+                RemoveActiveRunspace(runspace);
+            }
+
+            // Create and initialize a new Runspace
+            var iss = InitialSessionState.CreateDefault2();
+            iss.LanguageMode = (SystemPolicy.GetSystemLockdownPolicy() == SystemEnforcementMode.Enforce)
+                ? PSLanguageMode.ConstrainedLanguage : PSLanguageMode.FullLanguage;
+            runspace = RunspaceFactory.CreateRunspace(iss);
+            runspace.Name = runspaceName;
+            runspace.Open();
+            _activeRunspaces.TryAdd(runspace.Id, runspace);
+
+            return runspace;
+        }
+
+        private void ReturnRunspace(PSTaskBase task)
+        {
+            try
+            {
+                var runspace = task.Runspace;
+                Dbg.Assert(runspace != null, "Task runspace cannot be null.");
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    runspace.ResetRunspaceState();
+                    _runspacePool.Enqueue(runspace);
+                }
+                else
+                {
+                    RemoveActiveRunspace(runspace);
+                }
+            }
+            catch
+            {
```

Comment:
at least a comment on why we do nothing in the catch

Reply:
Will do.